scripts/generate-embeddings.ts

```
@@ -0,0 +1,426 @@
+      } else if (stats.isFile()) {
+        return [
+          {
+            path: path,
+            parentPath,
+          },
+        ];
+      } else {
+        return [];
+      }
+    })
+  );
+
+  const flattenedFiles = recursiveFiles.reduce(
+    (all, folderContents) => all.concat(folderContents),
+    []
+  );
+
+  return flattenedFiles.sort((a, b) => a.path.localeCompare(b.path));
+}
+
+abstract class BaseEmbeddingSource {
+  checksum?: string;
+  sections?: Section[];
+
+  constructor(
+    public source: string,
+    public path: string,
+    public parentPath?: string
+  ) {}
+
+  abstract load(): Promise<{
+    checksum: string;
+    sections: Section[];
+  }>;
+}
+
+class MarkdownEmbeddingSource extends BaseEmbeddingSource {
+  type: 'markdown' = 'markdown';
+
+  constructor(
+    source: string,
+    public filePath: string,
+    public parentFilePath?: string
+  ) {
+    const path = filePath.replace(/^docs/, '').replace(/\.md?$/, '');
+    const parentPath = parentFilePath
+      ?.replace(/^docs/, '')
+      .replace(/\.md?$/, '');
+
+    super(source, path, parentPath);
+  }
+
+  async load() {
+    const contents = await readFile(this.filePath, 'utf8');
+
+    const { checksum, sections } = processMdxForSearch(contents);
+
+    this.checksum = checksum;
+    this.sections = sections;
+
+    return {
+      checksum,
+      sections,
+    };
+  }
+}
+
+type EmbeddingSource = MarkdownEmbeddingSource;
+
+async function generateEmbeddings() {
+  const argv = await yargs.option('refresh', {
+    alias: 'r',
+    description: 'Refresh data',
+    type: 'boolean',
+  }).argv;
+
+  const shouldRefresh = argv.refresh;
+
+  if (
+    !process.env.NX_NEXT_PUBLIC_SUPABASE_URL ||
+    !process.env.NX_SUPABASE_SERVICE_ROLE_KEY ||
+    !process.env.NX_OPENAI_KEY
+  ) {
+    return console.log(
+      'Environment variables NEXT_PUBLIC_SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY, and OPENAI_KEY are required: skipping embeddings generation'
+    );
+  }
```

Can we split this into 3 checks and throw errors with the message instead of returning console.log? `return console.log(...)` works technically but gives the appearance of returning a value when it is just returning void or undefined, I'd rather throw or explicitly have a `return;`